src/main/haskell/kore/src/Kore/Step/MultiAnd.hs

```diff
@@ -0,0 +1,136 @@
+{-|
+Module      : Kore.Step.MultiAnd
+Description : Data structures and functions for manipulating
+              And with any number of children.
+Copyright   : (c) Runtime Verification, 2019
+License     : NCSA
+Maintainer  : virgil.serbanuta@runtimeverification.com
+Stability   : experimental
+Portability : portable
+-}
+module Kore.Step.MultiAnd
+    ( MultiAnd
+    , extractPatterns
+    , make
+    ) where
+
+import           Control.DeepSeq
+                 ( NFData )
+import qualified Data.Set as Set
+import           GHC.Generics
+                 ( Generic )
+
+import Kore.TopBottom
+       ( TopBottom (..) )
+
+{-| 'MultiAnd' is a Matching logic and of its children
+
+-}
+{- TODO (virgil): Make 'getMultiAnd' a non-empty list ("Data.NonEmpty").
+
+An empty 'MultiAnd' corresponding to 'Top' actually discards information
+about the sort of its child patterns! That is a problem for simplification,
+which should preserve pattern sorts.
+
+A non-empty 'MultiAnd' would also have a nice symmetry between 'Top' and
+'Bottom' patterns.
+-}
+newtype MultiAnd child = MultiAnd { getMultiAnd :: [child] }
+  deriving
+    (Applicative, Eq, Foldable, Functor, Generic, Monad, Ord, Show, Traversable)
+
+instance NFData child => NFData (MultiAnd child)
+
+instance TopBottom child => TopBottom (MultiAnd child)
+  where
+    isTop (MultiAnd []) = True
+    isTop _ = False
+    isBottom (MultiAnd [child]) = isBottom child
+    isBottom _ = False
+
+{-| 'AndBool' is an some sort of Bool data type used when evaluating things
+inside an 'MultiAnd'.
+-}
+-- TODO(virgil): Refactor, this is the same as OrBool
+data AndBool = AndTrue | AndFalse | AndUnknown
+
+{-|Does a very simple attempt to check whether a pattern
+is top or bottom.
+-}
+-- TODO(virgil): Refactor, this is the same as patternToOrBool
+patternToAndBool
+    :: TopBottom term
+    => term -> AndBool
+patternToAndBool patt
+  | isTop patt = AndTrue
+  | isBottom patt = AndFalse
+  | otherwise = AndUnknown
+
+{-| 'make' constructs a normalized 'MultiAnd'.
+-}
+make
+    :: (Ord term, TopBottom term)
+    => [term]
+    -> MultiAnd term
+make patts = filterAnd (MultiAnd patts)
+
+{-| Returns the patterns inside an @\and@.
+-}
+extractPatterns
+    :: TopBottom term
+    => MultiAnd term
+    -> [term]
+extractPatterns = getMultiAnd
+
+
+{- | Simplify the conjunction.
+
+The arguments are simplified by filtering on @\\top@ and @\\bottom@. The
+idempotency property of disjunction (@\\and(φ,φ)=φ@) is applied to remove
+duplicated items from the result.
+
+See also: 'filterUnique'
+-}
+filterAnd
+    :: (Ord term, TopBottom term)
+    => MultiAnd term
+    -> MultiAnd term
+filterAnd =
+    filterGeneric patternToAndBool . filterUnique
+
+
+{- | Simplify the disjunction by eliminating duplicate elements.
+
+The idempotency property of disjunction (@\\or(φ,φ)=φ@) is applied to remove
```

Comment:
```suggestion
The idempotency property of conjunction (@\\and(φ,φ)=φ@) is applied to remove
```
Actually, I'm not sure if this should be done here or in the simplifier. Although it doesn't require the full machinery of the simplifier, removing duplicate terms is a matter of policy and not representation. Also, we can only apply this simplification to certain types of arguments. I guess the same comment would apply to `MultiOr` though; this is consistent.

Reply:
I agree that the design of this is suboptimal. This also applies to the `makePredicate*` functions which follow the same pattern. However, IIRC, this made testing easier. I don't know if this has any influence on execution, I would guess that it does not.